semeio/workflows/ahm_analysis/ahmanalysis.py

```diff
@@ -0,0 +1,456 @@
+    try:
+        mygrid = Grid(input_grid, fformat="eclipserun")
+        dataframeg = mygrid.dataframe(activeonly=False)
+        return dataframeg
+    except OSError as err:
+        raise OSError("A grid with .EGRID format is expected.") from err
+
+
+def calc_delta_grid(
+    all_input_post, all_input_prior, mygrid_ok, caseobs, mygrid_ok_short
+):
+    """calculate mean delta of field grid data"""
+    delta_post_prior = np.subtract(all_input_prior, all_input_post)
+    delta_post_prior = np.absolute(delta_post_prior)
+    # Can we avoid changing in place here?
+    mygrid_ok["Mean_delta"] = np.mean(delta_post_prior, axis=0)
+    df_mean_delta = mygrid_ok.groupby(["IX", "JY"])[["Mean_delta"]].mean().reset_index()
+    mygrid_ok_short["Mean_D_" + caseobs] = df_mean_delta["Mean_delta"]
+    return mygrid_ok_short
+
+
+def _get_parameter_keys(genkw_parameters, prior_keys):
+    """excluding duplicate transformed parameter keys"""
+    transformed_param_keys = []
+    for parameter in genkw_parameters:
+        for key in prior_keys:
+            if "_" + parameter + ":" in key:
+                transformed_param_keys.append(key)
+    return set(prior_keys).difference(transformed_param_keys)
+
+
+def get_updated_parameters(prior_data, parameters):
+    """make list of updated parameters
+    (excluding duplicate transformed parameters)
+    """
+    parameter_keys = _get_parameter_keys(parameters, prior_data.keys())
+    # remove parameters with constant prior distribution
+    p_keysf = []
+    for dkey in parameter_keys:
+        if prior_data[dkey].ndim > 1:
+            warnings.warn("WARNING: Parameter " + dkey + " defined several times.")
+            flatten_arr = np.ravel(prior_data[dkey])
+            result = np.all(prior_data[dkey] == flatten_arr[0])
+            if not result:
+                p_keysf.append(dkey)
+        else:
+            if not all(x == prior_data[dkey][0] for x in prior_data[dkey]):
+                p_keysf.append(dkey)
+    return p_keysf
+
+
+def calc_ks(ks_matrix, dkeys, prior_data, target_data, key):
+    """Calculate KS matrix"""
+    for dkey in sorted(dkeys):
+        ks_val = ks_2samp(prior_data[dkey], target_data[dkey])
+        # Same comment here, possible to avoid changing in place?
+        ks_matrix.at[dkey, key] = ks_val[0]
+    return ks_matrix
+
+
+def check_names(ert_currentname, prior_name, target_name):
+    """Check input names given"""
+    if prior_name is None:
+        prior_name = ert_currentname
+    if target_name == "<ANALYSIS_CASE_NAME>":
+        target_name = "analysis_case"
+    return prior_name, target_name
+
+
+def check_inputs(misfitdata, prior_data):
+    """Check input ensemble prior is not empty
+    and if ensemble contains parameters for hm"""
+    if misfitdata.empty:
+        raise ValidationError("Empty prior ensemble")
+    if prior_data.empty:
+        raise ValidationError("Empty parameters set for History Matching")
+    return prior_data
```

Comment:
Not sure if I get the reasoning behind this function. We simply pass along two dataframes (or any instance with an `empty` property), and if neither is `empty` we return the second argument.

In the `run` function, the `misfitdata` is loaded in the function call, but it is never used. All we do is check that it isn't empty, is that vital for this job to work? Is it merely a sanity check?

Reply:
If this is the intended behaviour though I suggest that we don't return (since it's the same argument value), and we are explicit about what we do:

```
def raise_if_empty(dataframes, messages):
    for df, msg in zip(dataframes, messages):
        if df.empty:
            raise ValidationError(f"{message}")
```

And then call with:

```
prior_data = GenKwCollector.loadAllGenKwData(ert, prior_name)
raise_if_empty(
    dataframes=[prior_data, MisfitCollector.loadAllMisfitData(ert, prior_name)],
    messages=[
        "Empty prior ensemble",
        "Empty parameters set for History Matching",
    ],
)
```

Reply:
suggestion ahs been implemented. The point of the checks is to not start calulcations if some inputs data are missing.